src/main/java/org/folio/service/pieces/PiecesClaimingService.java

```diff
@@ -0,0 +1,238 @@
+  }
+
+  private static ClaimingResults createEmptyClaimingResults(String message) {
+    return new ClaimingResults().withClaimingPieceResults(List.of(new ClaimingPieceResult().withError(new Error().withMessage(message))));
+  }
+
+  private List<ClaimingPieceResult> createSuccessClaimingResults(List<List<String>> updatedPieceLists) {
+    return updatedPieceLists.stream().flatMap(Collection::stream).distinct()
+        .map(pieceId -> new ClaimingPieceResult().withPieceId(pieceId).withStatus(SUCCESS))
+        .toList();
+  }
+
+  private List<ClaimingPieceResult> createErrorClaimingResults(Map<String, List<String>> pieceIdsByVendorId, String message) {
+    return pieceIdsByVendorId.values().stream()
+        .flatMap(Collection::stream)
+        .map(pieceId -> new ClaimingPieceResult().withPieceId(pieceId).withStatus(FAILURE).withError(new Error().withMessage(message)))
+        .toList();
+  }
+
+  private static boolean isExportTypeClaimsAndCorrectVendorId(String vendorId, Map.Entry<String, Object> pieceIdsByVendorIdEntry) {
+    return pieceIdsByVendorIdEntry.getKey().startsWith(String.format("%s_%s", EXPORT_TYPE_CLAIMS, vendorId));
+  }
+
+  private Future<List<String>> updatePiecesAndCreateJob(List<String> pieceIds, Map.Entry<String, Object> pieceIdsByVendorIdEntry,
+                                                        RequestContext requestContext) {
+    log.info("updatePiecesAndCreateJob:: Updating pieces and creating a job, job key: {}, count: {}", pieceIdsByVendorIdEntry.getKey(), pieceIds.size());
+    return pieceUpdateFlowManager.updatePiecesStatuses(pieceIds, PieceBatchStatusCollection.ReceivingStatus.CLAIM_SENT, requestContext)
+        .compose(v -> createJob(pieceIdsByVendorIdEntry.getKey(), pieceIdsByVendorIdEntry.getValue(), pieceIds, requestContext).map(pieceIds));
+  }
+
+  private Future<Void> createJob(String configKey, Object configValue, List<String> pieceIds, RequestContext requestContext) {
+    var integrationDetail = new JsonObject(String.valueOf(configValue));
+    integrationDetail.getJsonObject(EXPORT_TYPE_SPECIFIC_PARAMETERS.getValue())
+      .getJsonObject(VENDOR_EDI_ORDERS_EXPORT_CONFIG.getValue())
+      .put(CLAIM_PIECE_IDS.getValue(), pieceIds);
+    return restClient.post(resourcesPath(DATA_EXPORT_SPRING_CREATE_JOB), integrationDetail, Object.class, requestContext)
+      .compose(response -> {
+        var createdJob = new JsonObject(String.valueOf(response));
+        log.info("createJob:: Created job, config key: {}, job status: {}", configKey, createdJob.getString(JOB_STATUS));
+        return restClient.postEmptyResponse(resourcesPath(DATA_EXPORT_SPRING_EXECUTE_JOB), createdJob, requestContext)
+          .onSuccess(v -> log.info("createJob:: Executed job, config key: {}", configKey));
+      })
+      .onComplete(asyncResult -> {
+        if (asyncResult.failed()) {
+          throw new IllegalStateException(CANNOT_COMPLETE_REQ.getValue(), asyncResult.cause());
```

Comment:
thinking about this line, so our future will be ended before onComplete and client that invoked that future can get first exception, not illegalStateException